### ISSUE TITLE
[RFC] Multi-turn SFT

### ISSUE THREAD

**Issue by haydn-jones** [pytorch/torchtitan#2757]
Currently multi-turn SFT isn't supported, I assume the primary reason is because the code uses EOS tokens for detecting document boundaries, which won't work for tokenizers like Qwen3:

```py
In [6]: tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-0.6B")

In [7]: chat = [
   ...:     {"role": "user", "content": "Hi!"},
   ...:     {"role": "assistant", "content": "Hello there!"}
   ...: ]

In [8]: tokenizer.apply_chat_template(chat, tokenize=False)
Out[8]: '<|im_start|>user\nHi!<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\nHello there!<|im_end|>\n'

In [9]: tokenizer.eos_token
Out[9]: '<|im_end|>
```

Just allowing multi-turn as it stands would make the 2nd assistant turn into a separate document.

This can probably be fixed pretty simply: both dataset styles (HF and Chat) are emitting positions that reset to 0 at packed conversation boundaries. Mask builders can just look at position resets instead of scanning for EOS:
```python
positions[i] <= positions[i-1] 
```
means "new document here," and will work with tokenizers like Qwen that use EOS as end of turn.

Outline would be like:

- Mask builder (`attention.py`)
  - Add `get_document_mask_from_positions(positions)` for flex path
  - Add `create_varlen_metadata_from_positions(positions)` for varlen path
  - Delete `get_document_mask_mod` and `create_varlen_metadata_for_document`
- Decoder (decoder.py)
  - `get_attention_masks` / `_get_flex_attention_masks`: read `extra_inputs["positions"]` instead of `(input_batch, tokenizer.eos_id)`
- Model-specific (`gpt_oss/model.py`, `llama4/model.py`)
  - Same swap in their `get_attention_masks`
- ChatDataset (`text_datasets.py`)
  - `_validate_messages`: accept multi-turn (optional system, alternating user/assistant, ends with assistant)
  - New `_get_assistant_spans`: prefix re-tokenization loop returning [(start, end), ...] like what is already in there
  - `_tokenize_sample`: mask everything, unmask assistant spans
- Tests
  - Multi-turn masking correctness
  - Cross-turn attention within a conversation
  - Cross-conversation isolation
  - Regression against existing single-turn configs

I've been using my own multi turn chat implementation on top of torchtitan so getting rid of that on my end would be nice.

Not sure if anyone is already working on this (@joecummings maybe?). I'd be happy to try if the plan is reasonable.

**Comment by joecummings** [comment]
@haydn-jones This is a good outline to get started - happy to review your PR when ready.

**Comment by haydn-jones** [comment]
Lol, I think I've already hit a snag. `positions` does *not* only reset on document boundaries. In the normal text dataset it will reset if a document is longer than `seq_len`:

https://github.com/pytorch/torchtitan/blob/b01adfb544b4331ecab090ebdb50b2296cd8eb6a/torchtitan/hf_datasets/text_datasets.py#L133

I suppose I should consider shipping document boundaries out of the dataset instead of trying to re-use positions.

Edit: this RoPE position resetting may be causing strange behavior  on main (long document resets position because > `seq_len`, but no EOS because its one doc so attn mask spans the whole doc in packed settings, causing RoPE to reset in the middle of a document) but I'm a bit too tired to verify.

**Comment by joecummings** [comment]
Yes, for `HuggingFaceTextDataset` (continuous text stream), we allow "chunking" in which we split a doc across a sequence boundary. In this case, positions *do* reset across this boundary which allows us to not worry about a document that grows even too long for the RoPE cache. While this may not be ideal, a split document itself is unable to attend to the previous chunk so that information is lost regardless + RoPE encodes relative information (yes, I ack differences in frequencies short vs long positions) so it shouldn't affect the loss **too** much and I made this choice for simplicity sake and to keep consistent the notion that seq_len is THE length you're training your model to comprehend. Perhaps the absolute simplest fix would be to truncate long documents ...

Also to the point of your edit: ``get_document_mask_mod`` sets the last token in a sequence to `eos_id` even if document is split and eos isn't there. 

***

But for chat datasets re the multi-turn work:

* Avg len(document) << seq_len
* Prefer [drop](https://github.com/pytorch/torchtitan/blob/b01adfb544b4331ecab090ebdb50b2296cd8eb6a/torchtitan/hf_datasets/text_datasets.py#L346) (or push to next buffer) over truncate/chunk

Under these constraints, there will never be a doc in the buffer than spans over the `seq_len` boundary.

**Comment by haydn-jones** [comment]
Ok yeah, I’m with you that this is not an issue for the chat side. On the continuous-text side, I think the important behavior is that the position wrap can happen mid-row. Using the actual `HuggingFaceTextDataset` with a long pseudo-document and `seq_len=4`, I see:

```
row 0: input=[2, 671, 671, 671] positions=[0, 1, 2, 3] labels=[671, 671, 671, 671]
row 1: input=[671, 671, 671, 671] positions=[1, 2, 3, 0] labels=[671, 671, 671, 671]
row 2: input=[671, 671, 671, 671] positions=[2, 3, 0, 1] labels=[671, 671, 671, 671]
row 3: input=[671, 671, 671, 671] positions=[3, 0, 1, 2] labels=[671, 671, 671, 671]
row 4: input=[671, 671, 671, 671] positions=[0, 1, 2, 3] labels=[671, 671, 671, 671]
```
(the code for this is https://gist.github.com/haydn-jones/f33ead4e9147fab52c2db3e2710ad94c)

So on `main`, for `HuggingFaceTextDataset + block_causal (flex or varlen)`, explicit RoPE positions can wrap mid-row while the EOS-based mask still treats the whole row as one attention block unless there is a real EOS inside it. With my change, the wrap in `positions` would also induce an attention boundary, so a row like `[2, 3, 0, 1]` effectively becomes two attention chunks: `[2, 3]` and `[0, 1]`.

That seems more internally consistent with the current wrapped-position behavior for long continuous text, even if the chunk semantics are a little different from the old EOS-based mask.

Edit: this rolling is really an artifact of slicing the global token/position stream into `seq_len + 1` shifted chunks. If positions were recomputed per emitted chunk, each row could start at 0 and this specific issue would go away.

**Comment by joecummings** [comment]
Ah, I see what you mean. Yes your fix seems consistent - thanks for clarifying

**Comment by haydn-jones** [comment]
Ok, well I opened the PR but realized I'm not properly detecting the assistant boundaries to do assistant-only loss. This seems like a hard problem to solve. Progressively tokenizing the conversation to find boundary offsets doesn't work as Qwen3 will insert `<think>  </think>` during the progressive tokenization if it the current subset ends with an assistant message, which don't exist when tokenizing the full conversation.

This is an *easy* problem to solve if we rely on https://huggingface.co/docs/transformers/v5.4.0/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase.apply_chat_template.return_assistant_tokens_mask, but this is not included in all chat templates (like with GPT-OSS). Adding it manually is not difficult, so one option would be to allow specifying a custom chat template path and requiring the `{% generation %}` keywords.

Edit: HF's own SFT requires this for assistant-only loss I believe, maybe not unreasonable as a requirement here, with custom chat template support for models without the keywords.

**Comment by haydn-jones** [comment]
Alright, my conclusion is the only reasonable thing to do is require `{% generation %}` in the chat templates and use `apply_chat_template` from the tokenizer, or not support multi-turn. Seems like this became a hard requirement in HF exactly because detecting assistant spans using other methods was model dependent and brittle (https://github.com/huggingface/transformers/pull/30650#issuecomment-2283900585). I could of course write detectors for the models supported here individually, but I don't like that, and using `apply_chat_template` would make the dataset code much cleaner.

The issue is no model that torchtitan supports has this keyword. I would be more than happy to insert the keywords and place the chat-templates in the repository, but I'm not sure how people feel about that.

`¯\_(ツ)_/¯` @joecummings 

Edit: ok realistically I would write detectors for specific models if that's the only route because torchtitan supports a small set of models.

**Comment by rakkit** [comment]
thats why i suggest to ask user provide eos in early sft pr https://github.com/pytorch/torchtitan/pull/2148/changes

given most of models usaully have more than one actual EOS, we can just ask user to provide the eos they want to use,  for example of qwen3, its `<|endoftext|>`.   (in princple any of extended and unsed token can be eos for infer the boundary of document)

**Comment by haydn-jones** [comment]
My changes no longer rely on EOS for detecting document boundaries, so that is solved. What isn't solved is identifying the boundary of assistant messages.

**Comment by haydn-jones** [comment]
Well, I've implemented model specific assistant span detection and marked the PR as ready. I apologize for how all over the place this was, feel free to review whenever is convenient.
